assembly/index.ts

```diff
@@ -17,37 +17,33 @@ export function initState(seed: f64): void {
   s1 = splitmix64(s0);
 }
 
-// Random between 0 and 9007199254740991 inclusive.
-// 9007199254740991 == Number.MAX_SAFE_INTEGER
-@inline
-function nextUint53(): f64 {
-  return <f64>(nextUint64() >> 11);
-}
-
 // Random between 0 and 0.9999999999999999 inclusive.
 // This matches Math.random() >= 0 and < 1.
 // A uniform random between 0-9 would be `Math.floor(next() * 10)`
 // To be inclusive of 10 you would do `Math.floor(next() * 11)`. Don't
 // use `Math.round(next() * 10)` as it will split the probability between
 // 0 and 10.
 export function next(): f64 {
-  // MAX_SAFE_INTEGER == 9007199254740991
-  // 9007199254740992 would be 1.0
-  // 9007199254740991 / 9007199254740992 == 0.9999999999999999
-  // 2^-53 == 1 / 9007199254740992 == 1.1102230246251565e-16
-  return nextUint53() * 1.1102230246251565e-16;
+  return reinterpret<f64>(<u64>0x3FF << 52 | nextUint64() >> 12) - 1.0;
 }
 
 // Random u64 using xoroshiro128starstar
 // 2018 by David Blackman and Sebastiano Vigna (public domain)
 // http://xoshiro.di.unimi.it/xoroshiro128starstar.c
 @inline
 function nextUint64(): u64 {
-  let result: u64 = rotl<u64>(s0 * 5, 7) * 9;
+  let _s0 = s0;
+  let _s1 = s1;
+
+  let t = rotl((_s0 << 2) + _s0, 7);
```

Comment:
Please make this a separate PR, I don't know if this handles overflow the same as the original and I don't have tests yet to verify it behaves the same as the original, the only thing I have is to visually check the benchmark looks uniformly distributed at the end.

Reply:
Sure. done

Reply:
But transform `x * 5` to `(x << 2) + x` is safe operation and produce same overflow behaviour. The reason of doing this is most VMs can't optimize `mul/div/mod` by constant for 64-bit registers yet. You could see how this doing in GCC: https://godbolt.org/z/d15Ncm

Reply:
Breaking up a PR was a suggestion because I would have merged the other parts first while clarifying the above.  You've clarified it well enough that I'm good with it now.  Have you thought about adding the optimization to binaryen which assemblyscript uses to optimize?

Reply:
If this concerns arithmetic for constants then it is outscope of binaryen. Regarding optimization with cached global variables then it will be to some extent in the next release.